qa-tests-backend/src/test/groovy/ProcessesListeningOnPortsTest.groovy

```diff
@@ -0,0 +1,252 @@
+                    .setArgs(["(socat "+SOCAT_DEBUG+" TCP-LISTEN:80,fork STDOUT & " +
+                                      "socat "+SOCAT_DEBUG+" TCP-LISTEN:8080,fork STDOUT)" as String,]),
+            new Deployment()
+                    .setName(TCPCONNECTIONTARGET2)
+                    .setImage("quay.io/rhacs-eng/qa:socat")
+                    .addPort(8081, "TCP")
+                    .addLabel("app", TCPCONNECTIONTARGET2)
+                    .setExposeAsService(true)
+                    .setCommand(["/bin/sh", "-c",])
+                    .setArgs(["(socat "+SOCAT_DEBUG+" TCP-LISTEN:8081,fork STDOUT)" as String,]),
+        ]
+    }
+
+    def createDeployments() {
+        targetDeployments = buildTargetDeployments()
+        orchestrator.batchCreateDeployments(targetDeployments)
+        for (Deployment d : targetDeployments) {
+            assert Services.waitForDeployment(d)
+        }
+    }
+
+    def setupSpec() {
+        orchestrator.createNamespace(OTHER_NAMESPACE)
+        orchestrator.createImagePullSecret(
+                "quay",
+                Env.mustGet("REGISTRY_USERNAME"),
+                Env.mustGet("REGISTRY_PASSWORD"),
+                OTHER_NAMESPACE,
+                "https://quay.io"
+        )
+        orchestrator.createServiceAccount(
+                new K8sServiceAccount(
+                        name: "default",
+                        namespace: OTHER_NAMESPACE,
+                        imagePullSecrets: ["quay"]
+                )
+        )
+
+        createDeployments()
+    }
+
+    def destroyDeployments() {
+        for (Deployment deployment : targetDeployments) {
+            orchestrator.deleteDeployment(deployment)
+        }
+        for (Deployment deployment : targetDeployments) {
+            if (deployment.exposeAsService) {
+                orchestrator.waitForServiceDeletion(new Service(deployment.name, deployment.namespace))
+            }
+        }
+        orchestrator.deleteNamespace(OTHER_NAMESPACE)
+        orchestrator.waitForNamespaceDeletion(OTHER_NAMESPACE)
+    }
+
+    def cleanupSpec() {
+        destroyDeployments()
+    }
+
+    def rebuildForRetries() {
+        if (Helpers.getAttemptCount() > 1) {
+            log.info ">>>> Recreating test deployments prior to retest <<<<<"
+            destroyDeployments()
+            sleep(5000)
+            createDeployments()
+            sleep(5000)
+            log.info ">>>> Done <<<<<"
+        }
+    }
+
+    @Tag("BAT")
+    def "Verify networking endpoints with processes appear in API at the deployment level"() {
+        given:
+        "Two deployments that listen on ports are started up"
+
+        rebuildForRetries()
+        def clusterId = ClusterService.getClusterId()
+
+        String deploymentId1 = targetDeployments[0].getDeploymentUid()
+        String deploymentId2 = targetDeployments[1].getDeploymentUid()
+
+        def gotCorrectNumElements = waitForResponseToHaveNumElements(2, deploymentId1, 240)
```

Comment:
More of an observation, and not a comment blocking this PR:
Looking through the test logs, the first attempt fails after a couple of minutes, and the second attempt immediately succeeds after few seconds.. any idea what's going on there?

Reply:
If there is a failure, the retry should be independent. It does seem like the first try succeeds less often than the second try. We will investigate. We do have planned improvements that should make this more reliable.